buffer/src/main/java/io/netty/buffer/SizeClasses.java

```diff
@@ -0,0 +1,403 @@
+        int nDelta = 0;
+        while (nDelta < ndeltaLimit) {
+            size = sizeClass(index++, log2Group, log2Delta, nDelta++);
+        }
+        log2Group += LOG2_SIZE_CLASS_GROUP;
+
+        //All remaining groups, nDelta start at 1.
+        while (size < chunkSize) {
+            nDelta = 1;
+
+            while (nDelta <= ndeltaLimit && size < chunkSize) {
+                size = sizeClass(index++, log2Group, log2Delta, nDelta++);
+                normalMaxSize = size;
+            }
+
+            log2Group++;
+            log2Delta++;
+        }
+
+        //chunkSize must be normalMaxSize
+        assert chunkSize == normalMaxSize;
+
+        nSizes = index;
+    }
+
+    //calculate size class
+    private int sizeClass(int index, int log2Group, int log2Delta, int nDelta) {
+        short isMultiPageSize;
+        if (log2Delta >= pageShifts) {
+            isMultiPageSize = yes;
+        } else {
+            int pageSize = 1 << pageShifts;
+            int size = (1 << log2Group) + (1 << log2Delta) * nDelta;
+
+            isMultiPageSize = size == size / pageSize * pageSize? yes : no;
+        }
+
+        int log2Ndelta = nDelta == 0? 0 : log2(nDelta);
+
+        byte remove = 1 << log2Ndelta < nDelta? yes : no;
+
+        int log2Size = log2Delta + log2Ndelta == log2Group? log2Group + 1 : log2Group;
+        if (log2Size == log2Group) {
+            remove = yes;
+        }
+
+        short isSubpage = log2Size < pageShifts + LOG2_SIZE_CLASS_GROUP? yes : no;
+
+        int log2DeltaLookup = log2Size < LOG2_MAX_LOOKUP_SIZE ||
+                              log2Size == LOG2_MAX_LOOKUP_SIZE && remove == no
+                ? log2Delta : no;
+
+        short[] sz = {
+                (short) index, (short) log2Group, (short) log2Delta,
+                (short) nDelta, isMultiPageSize, isSubpage, (short) log2DeltaLookup
+        };
+
+        sizeClasses[index] = sz;
+        int size = (1 << log2Group) + (nDelta << log2Delta);
+
+        if (sz[PAGESIZE_IDX] == yes) {
+            nPSizes++;
+        }
+        if (sz[SUBPAGE_IDX] == yes) {
+            nSubpages++;
+            smallMaxSizeIdx = index;
+        }
+        if (sz[LOG2_DELTA_LOOKUP_IDX] != no) {
+            lookupMaxSize = size;
+        }
+        return size;
+    }
+
+    private void idx2SizeTab() {
+        sizeIdx2sizeTab = new int[nSizes];
+        pageIdx2sizeTab = new int[nPSizes];
+
+        int pageIdx = 0;
+
+        for (int i = 0; i < nSizes; i++) {
+            short[] sizeClass = sizeClasses[i];
+            int log2Group = sizeClass[LOG2GROUP_IDX];
+            int log2Delta = sizeClass[LOG2DELTA_IDX];
+            int nDelta = sizeClass[NDELTA_IDX];
+
+            int size = (1 << log2Group) + (nDelta << log2Delta);
+            sizeIdx2sizeTab[i] = size;
+
+            if (sizeClass[PAGESIZE_IDX] == yes) {
+                pageIdx2sizeTab[pageIdx++] = size;
+            }
+        }
+    }
+
+    private void size2idxTab() {
```

Comment:
we should merge this code into the constructor to be able to define `size2idxTab` as final